server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceInput.java

```diff
@@ -9,14 +9,55 @@
 package org.elasticsearch.cluster.routing.allocation.allocator;
 
 import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
 import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
 
-import java.util.List;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toUnmodifiableSet;
 
 /**
  * The input to the desired balance computation.
  *
  * @param routingAllocation a copy of (the immutable parts of) the context for the allocation decision process
  * @param ignoredShards     a list of the shards for which earlier allocators have claimed responsibility
  */
-public record DesiredBalanceInput(long index, RoutingAllocation routingAllocation, List<ShardRouting> ignoredShards) {}
+public record DesiredBalanceInput(long index, RoutingAllocation routingAllocation, Set<ShardRouting> ignoredShards) {
+
+    public static DesiredBalanceInput create(long index, RoutingAllocation routingAllocation) {
+        return new DesiredBalanceInput(
+            index,
+            routingAllocation.immutableClone(),
+            getIgnoredShardsWithDiscardedAllocationStatus(routingAllocation)
+        );
+    }
+
+    private static Set<ShardRouting> getIgnoredShardsWithDiscardedAllocationStatus(RoutingAllocation routingAllocation) {
+        return routingAllocation.routingNodes()
+            .unassigned()
+            .ignored()
+            .stream()
+            .map(
+                shardRouting -> shardRouting.updateUnassigned(
+                    discardAllocationStatus(shardRouting.unassignedInfo()),
+                    shardRouting.recoverySource()
+                )
+            )
+            .collect(toUnmodifiableSet());
```

Comment:
I wondered if this might cause us problems related to nondeterministic ordering. It doesn't look like it does (today at least) and previously we copied this into a `HashSet` anyway. Therefore 👍  but writing this comment for the avoidance of doubt.